### PR TITLE
Attempting infinite train data

### PR DESCRIPTION
**Pull request by srbhchandra**
Objective of this PR: recreate and use the synthetic train dataset on the fly during training.
This is achieved by recreating the `iam_syn_paragraphs.data_train` in `iam_original_and_synthetic_paragraphs.train_dataloader()` and instantiating the training using the command line arg `--reload_dataloaders_every_n_epochs=n`.

We set n to 40 in our experiments.

Wandb run - https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/runs/1u8h1c7k
We get our best metrics on `test` so far:
```
cer: 0.1250961422920227
loss: 0.7417508363723755
```
Wandb report - https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/reports/Report-for-misunderstood-galaxy-5--VmlldzoyMzgwODg5

### Misc

Past Wandb run - https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/runs/s9d2pgj4
This didn't train well (`validation loss` kept increasing after a few epochs) because of 1 bug: we didn't set `IAMSyntheticParagraphs.transform.scale_factor = 1` which was needed since we are using the IAM synthetic paragraphs validation dataset. So `IAMSyntheticParagraphs.transform.scale_factor` defaulted to 2.
We reverted to not using the IAM synthetic paragraphs validation dataset but also started setting `IAMSyntheticParagraphs.transform.scale_factor = 1`.

Signed-off-by: srbhchandra <srbhchandra@gmail.com>

### CODE REVIEW

**Comment by charlesfrye** [comment]
fyi, if you want `pre-commit` to run before your commits so you don't fail the checks on the PR,
[just run `pre-commit install`](https://pre-commit.com/#3-install-the-git-hook-scripts) while inside of the repo and it should get added to your git hooks. pre-commit will already be installed via the `dev` requirements

**Comment by charlesfrye** [comment]
there's an open issue to improve the CONTRIBUTING.md so that this is included.

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 1]
```diff
@@ -56,11 +58,14 @@ def prepare_data(self, *args, **kwargs) -> None:
...
         rank_zero_info(f"IAMSyntheticParagraphs.setup({stage}): Loading trainval IAM paragraph regions and lines...")
 
-        if stage == "fit" or stage is None:
-            line_crops, line_labels = load_line_crops_and_labels("train", PROCESSED_DATA_DIRNAME)
+        def _load_dataset(split: str, transform: ParagraphStem) -> BaseDataset:
+            line_crops, line_labels = load_line_crops_and_labels(split, PROCESSED_DATA_DIRNAME)
             X, para_labels = generate_synthetic_paragraphs(line_crops=line_crops, line_labels=line_labels)
             Y = convert_strings_to_labels(strings=para_labels, mapping=self.inverse_mapping, length=self.output_dims[0])
-            self.data_train = BaseDataset(X, Y, transform=self.trainval_transform)
+            return BaseDataset(X, Y, transform=transform)
```
Comment (srbhchandra) [inline text_recognizer/data/iam_synthetic_paragraphs.py:65]: keeping this reformating so we can easily add `self.data_val` if we want. 
In fact, I did add it to a run but a different bug made the run unsuccessful.

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 2]
```diff
@@ -29,6 +30,7 @@ class IAMSyntheticParagraphs(IAMParagraphs):
     def __init__(self, args: argparse.Namespace = None):
         super().__init__(args)
         self.trainval_transform.scale_factor = 1  # we perform rescaling ahead of time, in prepare_data
+        self.transform.scale_factor = 1
```
Comment (srbhchandra) [inline text_recognizer/data/iam_synthetic_paragraphs.py:33]: Not having this line is the bug discussed in the below comment.

File: text_recognizer/data/iam_original_and_synthetic_paragraphs.py [hunk 3]
```diff
@@ -39,9 +39,22 @@ def setup(self, stage: str = None) -> None:
         if stage == "fit" or stage is None:
             self.data_train = ConcatDataset([self.iam_paragraphs.data_train, self.iam_syn_paragraphs.data_train])
             self.data_val = self.iam_paragraphs.data_val
+
         if stage == "test" or stage is None:
             self.data_test = self.iam_paragraphs.data_test
 
+    def train_dataloader(self):
```
Comment (charlesfrye) [inline text_recognizer/data/iam_original_and_synthetic_paragraphs.py:46]: I think putting the data generation inside this method is a neat way to quickly get "infinite data" with Lightning. But I think we need to be careful not to break assumptions/promises of the Lightning API.

1. We're calling the `.setup` method -- that's part of the Lightning API, with expectations and guarantees about when it's called. We don't want a future contributor (e.g. us in six months) to expect it to conform to that API and end up burned. I think this is easily resolvable though. Elsewhere, I suggested refactoring out `_load_dataset` from `.setup` and making it its own method. We could call that method both here and in `.setup`.

2. For the first epoch and in the case where we never do a dataloader reload, it seems like we're needlessly generating the data twice: first when `setup` gets called and then again when `train_dataloader` gets called. It should only be generated once. That should be done in `.setup`, since it's expected that `.setup` does everything that's needed to prep a dataset. We could set a flag when `.setup` gets called 

The other way to do "infinite data" would be to incorporate the image synthesis into the `Dataset` itself. This would be a new class that lives in `iam_synthetic_paragraphs`, inheriting from the `torch.utils.data.Dataset`.  You'd use the `__getitem__` method to call something like `generate_synthetic_paragraphs`, but working one paragraph at a time. You'll want to handle `transforms` the same way as `BaseDataset`.

There's a potential for slowdown, but I suspect our batch size limitations from GPU RAM are enough that multiprocessing `DataLoader`s can generally serve up a full batch by the time the GPU finishes a fwd/bwd pass. You can use the `--profile` flag (with `--gpus=1, `--max_epochs=1` and the same `--num_workers` as in the `ddp` case) to check the detailed trace or just use the GPU utilization reported to W&B by `nvidia-smi` as a spot check (should be 70%+). 

See StackOverflow discussion of infinite datasets in PyTorch [here](https://stackoverflow.com/questions/54359243/implementing-an-infinite-loop-dataset-dataloader-in-pytorch) and watch out for the RNG issues they mention!

Note that `__len__` can be set to return whatever number you wish as needed, e.g. by a command line argument. 

FYI for 2, I think the pseudocode version of what Lightning is doing with the `DataModule` is something like:
```python
# download, etc...
dm = TheDataModule()
dm.prepare_data()

# splits/transforms
dm.setup(stage="fit")

if not reload_dataloaders_every_n_epochs:
    train_loader = dm.train_dataloader()
# fitting loop
for epoch in epochs:
  if reload_dataloaders_every_n_epochs and not epoch % reload_dataloaders_every_n_epochs:
        train_loader = dm.train_dataloader()

  for batch in dm.train_dataloader():
      ...
  
  for batch in dm.val_dataloader():
      ...

dm.teardown(stage="fit")

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 4]
```diff
@@ -56,11 +58,14 @@ def prepare_data(self, *args, **kwargs) -> None:
     def setup(self, stage: str = None) -> None:
         rank_zero_info(f"IAMSyntheticParagraphs.setup({stage}): Loading trainval IAM paragraph regions and lines...")
 
-        if stage == "fit" or stage is None:
-            line_crops, line_labels = load_line_crops_and_labels("train", PROCESSED_DATA_DIRNAME)
+        def _load_dataset(split: str, transform: ParagraphStem) -> BaseDataset:
```
Comment (charlesfrye) [inline text_recognizer/data/iam_synthetic_paragraphs.py:61]: Is this type annotation too strict? Do we really only allow `ParagraphStem`s here? I don't think we actually have a useful type annotation to provide here, because the type is something like "callable that takes a tensor and returns a tensor", which is insufficient to express actually useful things like shape assumptions. So my preference would be not to annotate.

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 )
 from text_recognizer.data.util import BaseDataset, convert_strings_to_labels, resize_image
 import text_recognizer.metadata.iam_synthetic_paragraphs as metadata
+from text_recognizer.stems.paragraph import ParagraphStem
```
Comment (charlesfrye) [inline text_recognizer/data/iam_synthetic_paragraphs.py:19]: Heads up, I [refactored the `stems` a bit](https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/tree/main/text_recognizer/stems) a few days ago. Among the reasons to maybe try rebasing this one right before merge.

**Review by charlesfrye — CHANGES_REQUESTED** [review]
Nicely done getting infinite data working! Glad to see it improves our metrics.

I've got some thoughts on the design here. Let's do a pair session so we can exchange ideas faster and get this over the finish line. DMing you!

File: text_recognizer/data/iam_synthetic_paragraphs.py (hunk 1 above)
Comment (srbhchandra) [reply text_recognizer/data/iam_synthetic_paragraphs.py:65]: done at https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/pull/31/files#diff-cd205268606d2b02590586c359105225d4cffe942d0e76f9e4c50d33dd260e17R76

File: text_recognizer/data/iam_synthetic_paragraphs.py (hunk 2 above)
Comment (srbhchandra) [reply text_recognizer/data/iam_synthetic_paragraphs.py:33]: - [ ] I still need to handle resizing cleanly. I have cleaned up the IAM image loading and inversion logic.

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 6]
```diff
@@ -56,11 +58,14 @@ def prepare_data(self, *args, **kwargs) -> None:
...
-        if stage == "fit" or stage is None:
-            line_crops, line_labels = load_line_crops_and_labels("train", PROCESSED_DATA_DIRNAME)
+        def _load_dataset(split: str, transform: ParagraphStem) -> BaseDataset:
+            line_crops, line_labels = load_line_crops_and_labels(split, PROCESSED_DATA_DIRNAME)
             X, para_labels = generate_synthetic_paragraphs(line_crops=line_crops, line_labels=line_labels)
             Y = convert_strings_to_labels(strings=para_labels, mapping=self.inverse_mapping, length=self.output_dims[0])
-            self.data_train = BaseDataset(X, Y, transform=self.trainval_transform)
+            return BaseDataset(X, Y, transform=transform)
+
+        if stage in ["fit", "train_only", None]:
```
[the remark this answers is not in the document]
Comment (srbhchandra) [reply text_recognizer/data/iam_synthetic_paragraphs.py:67]: Done. In PR https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/pull/31 I removed "train_only"

File: text_recognizer/data/iam_synthetic_paragraphs.py (hunk 4 above)
Comment (srbhchandra) [reply text_recognizer/data/iam_synthetic_paragraphs.py:61]: Fixed at https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/pull/31/files#diff-cd205268606d2b02590586c359105225d4cffe942d0e76f9e4c50d33dd260e17R76

File: text_recognizer/data/iam_original_and_synthetic_paragraphs.py (hunk 3 above)
Comment (srbhchandra) [reply text_recognizer/data/iam_original_and_synthetic_paragraphs.py:46]: Check out https://github.com/full-stack-deep-learning/fsdl-text-recognizer-2022/pull/40

**Comment by charlesfrye** [comment]
see #41, which incorporates this PR
